synapse/lib/base.py

```diff
@@ -0,0 +1,571 @@
+import gc
+import atexit
+import signal
+import asyncio
+import logging
+import functools
+import threading
+import collections
+
+import synapse.exc as s_exc
+import synapse.glob as s_glob
+import synapse.common as s_common
+
+import synapse.lib.thishost as s_thishost
+
+logger = logging.getLogger(__name__)
+
+def _fini_atexit(): # pragma: no cover
+
+    for item in gc.get_objects():
+
+        if not isinstance(item, Base):
+            continue
+
+        if item.isfini:
+            continue
+
+        if not item._fini_atexit:
+            continue
+
+        try:
+            rv = item.fini()
+            if asyncio.iscoroutine(rv):
+                # Try to run the fini on its loop
+                loop = item.loop
+                if not loop.is_running():
+                    continue
+                loop.create_task(rv)
+
+        except Exception as e:
+            logger.exception('atexit fini fail: %r' % (item,))
+
+atexit.register(_fini_atexit)
+
+class Base:
+    '''
+    Base class for Synapse objects.
+
+    Acts as an observable, enables async init and fini.
+
+    Example:
+
+        class Foo(Anit):
+
+            def __init__(self, x):
+                self.x = x
+
+            async def __anit__(self):
+                await stuff()
+
+        foo = await Foo.anit(10)
+    '''
+    @classmethod
+    async def anit(cls, *args, **kwargs):
+        self = cls(*args, **kwargs)
+        self.loop = asyncio.get_event_loop()
+        await self.__anit__()
+        return self
+
+    def __init__(self):
+        self.isfini = False
```

Comment:
Can we set `self.loop = None` in `__init__`?  We can type-hint him as an ioloop.  I'm wary of defining instance vars outside of `__init__`.

Reply:
ok